### PR TITLE
Changes made for PhaseSystem to work with the Waxing/Waning

### PR DESCRIPTION
**Pull request by WispSS13**
Changes made on 18/11/2020 to allow us to better use the waxing/waning mechanic with more cards and relics,

### CODE REVIEW

File: src/main/java/org/m14n/moonthespire/PhaseSystem.java [hunk 1]
```diff
@@ -57,6 +69,31 @@ private static void modifyDexterity(int amt) {
...
+    public static void subscribe(PhaseChangeSubscriber subscriber) {
+        subscribers.add(subscriber);
+    }
+
+    public static void unsubscribe(PhaseChangeSubscriber subscriber) {
+        subscribers.remove(subscriber);
+    }
+
+
     static void reset() {
```
Comment (strangeglyph) [inline src/main/java/org/m14n/moonthespire/PhaseSystem.java:97]: Oops, we missed this when we were updating PhaseSystem. Could you reset wasWaxing and wasWaning too, here?

File: src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+import static org.m14n.moonthespire.Mod.cardPath;
+import static org.m14n.moonthespire.Mod.makeId;
+
+public class EbbAndFlow extends AbstractMoonCard implements PhaseChangeSubscriber {
+    public static final String ID = makeId("EbbAndFlow");
+    public static final String IMG = cardPath("moondrop.png");
+
+    private static final int COST = 2;
+    private static final CardType TYPE = CardType.ATTACK;
```
Comment (strangeglyph) [inline src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java:34]: Should we change the card type alongside the target in onPhaseChange?

File: src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    private static final CardColor COLOR = MoonCharacter.Enums.MOON_PURPLE;
+    private static final CardRarity RARITY = CardRarity.UNCOMMON;
+    private static final CardTarget TARGET = CardTarget.ENEMY;
+    private static final CardStrings cardStrings;
```
Comment (strangeglyph) [inline src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java]: Don't need this

File: src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    public EbbAndFlow() {
+        super(ID, IMG, COST, TYPE, COLOR, RARITY, TARGET);
+        PhaseSystem.subscribe(this);
```
Comment (strangeglyph) [inline src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java]: Style-wise I'd say move the subscription below the initialization of the members.

File: src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public void onRemoveFromMasterDeck() {
+        PhaseSystem.unsubscribe(this);
+    }
+
+    @Override
+    public void use(AbstractPlayer p, AbstractMonster m) {
+        // Create an int which equals to your current energy.
```
Comment (strangeglyph) [inline src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java]: Kind of superfluous comment (it's also wrong, it's not energy, it's hand size)

File: src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        this.addToTop(new DiscardAction(AbstractDungeon.player, AbstractDungeon.player, count, false));
+    }
+
+    @Override
+    public void upgrade() {
+        if (!this.upgraded) {
+            this.upgradeName();
+            this.exhaust = false;
+            this.upgradeDamage(2);
+            this.rawDescription = cardStrings.UPGRADE_DESCRIPTION;
```
Comment (strangeglyph) [inline src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java]: Unnecessary

File: src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+
+    public AbstractCard makeCopy() {
+        return new EbbAndFlow();
+    }
+
+    static {
+        cardStrings = CardCrawlGame.languagePack.getCardStrings("EbbAndFlow");
+    }
```
Comment (strangeglyph) [inline src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java]: Also unnecessary

**Review by strangeglyph — CHANGES_REQUESTED** [review]

File: src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java (hunk 6 above)
Comment (strangeglyph) [reply src/main/java/org/m14n/moonthespire/cards/EbbAndFlow.java]: To clarify, setting raw description is unnecessary. InitializeDescription does the work for us already

File: src/main/java/org/m14n/moonthespire/PhaseSystem.java (hunk 1 above)
Comment (WispSS13) [reply src/main/java/org/m14n/moonthespire/PhaseSystem.java:97]: Sure thing, wasWaxing and wasWaning now reset to False.

**Pull request merged**